notebook/notebook/notebook.py

```diff
@@ -113,201 +143,334 @@ def start_pod(jupyter_token, image):
+            maxCond = condition
+            maxDate = condition.last_transition_time
+
+    return {
+        "message": maxCond.message,
+        "reason": maxCond.reason,
+        "status": maxCond.status,
+        "type": maxCond.type}
+
+
+common_pod_paths = [
+    ['name', 'metadata.labels.name'],
+    ['pod_name', 'metadata.name'],
+    ['svc_name', 'metadata.labels.svc_name'],
+    ['pod_status', 'status.phase'],
+    ['creation_date', 'metadata.creation_timestamp',
+        lambda x: x.strftime('%D')],
+    ['token', 'metadata.labels.jupyter_token'],
+    ['container_status', 'status.container_statuses',
+        lambda x: read_containers_status(x)],
+    ['condition', 'status.conditions', lambda x: read_conditions(x)]
+]
+
+pod_paths = [
+    *common_pod_paths,
+    ['svc_status', 'metadata.labels.svc_name', lambda x: read_svc_status(x)],
+]
+
+pod_paths_post = [
+    *common_pod_paths,
+    ['svc_status', 'metadata.labels.svc_name', lambda _: 'Running'],
+]
+
+
+# Routes #
+
+
+# TODO: Simply checking ws.close isn't enough
+# https://github.com/heroku-python/flask-sockets/issues/60
+# Solution may be move to aiohttp
+@sockets.route('/jupyter/ws')
+def echo_socket(ws):
+    user_id = ws.environ['HTTP_USER']
+    w = kube.watch.Watch()
+
+    for event in w.stream(k8s.list_namespaced_pod, namespace='default',
+                          label_selector=f"user_id={UNSAFE_user_id_transform(user_id)}"):
+
+        if ws.closed:
+            log.info("Websocket closed")
+            w.stop()
+            return
+
+        try:
+            obj = event["object"]
+
+            if event['type'] == 'MODIFIED' or event['type'] == 'ADDED':
+                ws.send(
+                    f'{{"event": "{event["type"]}", "resource": {marshall_json([obj], pod_paths, True)}}}')
+
+            if event['type'] == 'DELETED':
+                ws.send(
+                    f'{{"event": "DELETED", "resource": {marshall_json([obj], pod_paths, True)}}}')
+
+        except Exception as e:
+            log.info("Issue with watcher")
+            log.error(e)
+            w.stop()
+            break
+
+    ws.close()
+
+
+@app.route('/jupyter/verify/<svc_name>/', methods=['GET'])
+def verify(svc_name: str):
+    access_token = request.cookies.get('access_token')
+
+    if is_falsy_str(access_token):
+        return '', 401
+
+    resp = requests.get(f'{AUTH_GATEWAY}/verify',
+                        headers={'Authorization': f'Bearer {access_token}'})
+
+    if resp.status_code != 200:
+        return '', 401
+
+    user_id = resp.headers.get('User')
+
+    if is_falsy_str(user_id):
+        return '', 401
+
+    k_res = k8s.read_namespaced_service(svc_name, 'default')
+    labels = k_res.metadata.labels
+
+    if labels['user_id'] != UNSAFE_user_id_transform(user_id):
+        return '', 401
+
+    # FIXME: Replace with pod ip
+    # resp = Response('')
+    # resp.headers['IP'] = k_res.spec.cluster_ip
```

Comment:
This will be modified and uncommented when I finish the move from services. This is what I settled on for passing IP to nginx for the redirect. I found that more elegant than exposing the IP directly to the web.